server/src-lib/Hasura/Server/Init.hs

```diff
@@ -116,6 +116,12 @@ data ServeOptions impl
   , soPlanCacheOptions :: !E.PlanCacheOptions
   }
 
+data DowngradeOptions
+  = DowngradeOptions
+  { dgoTargetVersion :: !T.Text
```

Comment:
The fact that this is a catalog version rather than a version of `graphql-engine` is a little confusing, since it means users would need to know about catalog versions, and passing a version of `graphql-engine` itself would result in an error message stating that the given version is unknown.

I don’t know exactly what the right thing to do is here, but one thing we could do is try to maintain a mapping between released versions of `graphql-engine` and catalog versions. Another thing we could do is keep the current interface and either document the mapping or just tell people to look in the error message when an older version of `graphql-engine` starts up (which lists which catalog version it expects).

@rikinsk, do you have any thoughts about this?

Reply:
I agree. The catalog version is something users never come across unless they need to downgrade. It just seems to make more sense for us to maintain the HGE version to catalog version mapping internally and the users can just specify the HGE version to downgrade to. It also doesnt seem to make any logical sense to move to an intermediate catalog version that doesnt correspond to an HGE version anyway IMO _(could be wrong here)_.

Reply:
For something that's going to be used so rarely, and presumably mostly at our instruction on Discord or whatever, it doesn't seem to be worth it to add a mapping that we then need to maintain. Is there a use case for this other than for support? In any case, if there is a mapping, I don't think it needs to exist in code, it could exist as documentation only.

Reply:
I’ll defer to @rikinsk’s judgment here, since he is the keeper of DX. I see the merits to both arguments. I agree that it’s awkward to have to maintain the mapping, and that it’s probably not hugely important, but OTOH if we’re going to maintain *documentation* about the mapping, I don’t really see why it would be any more difficult to maintain it in code instead (which would eliminate the need to maintain it in the docs).

The way I see the `downgrade` command is that it offers our users a sense of safety. It provides a safe way for them to get back to a known good state if they ever upgrade and discover that something breaks. Obviously we’d like it to be necessary as infrequently as possible, but I also think there is value in making the process as smooth and effortless as possible, as it will be used in a situation where someone is likely already feeling uneasy, and it doesn’t seem like too much effort to me. But yes, it’s a small detail, and I certainly won’t die on this hill.

Reply:
Note, we already have this mapping in documentation [here](https://github.com/hasura/graphql-engine/blob/a5c5609abbf704b8e4439b5d6270ef87271a6207/docs/graphql/manual/deployment/downgrading.rst). I'll replicate the mapping in the `optparse-applicative` code.

Reply:
I think we can (and should!) remove the mapping from the documentation as part of this change.

Reply:
Do 28, 29 and 30 all map to 1.0.0?

Reply:
According to the release tags, [1.0.0 was catalog version 28](https://github.com/hasura/graphql-engine/blob/v1.0.0/server/src-lib/Hasura/Server/Migrate/Version.hs), [1.0.0-beta.1 was version 29](https://github.com/hasura/graphql-engine/blob/v1.0.0-beta.1/server/src-lib/Hasura/Server/Migrate/Version.hs), and [1.0.0-beta.2 was version 30](https://github.com/hasura/graphql-engine/blob/v1.0.0-beta.2/server/src-lib/Hasura/Server/Migrate/Version.hs).